### PR TITLE
Re #2920: Fix memory leaks when replacing data

### PR DESCRIPTION
**Pull request by krisdages**
https://github.com/vega/vega/issues/2920

* When detaching Operators, clear the references to `source` and `pulse`
* When detaching EventStreams, reset the `_filter` function and clear `targets`
* When cleaning Facet, remove subflows with count 0 from the active _targets array.

These may be a little too much of a "blunt object" approach for the general use case, just wanted to share what fixed the issue for me.

### CODE REVIEW

File: packages/vega-dataflow/src/EventStream.js [hunk 1]
```diff
@@ -122,7 +122,10 @@ EventStream.prototype = {
   },
 
   detach() {
-    // no-op for handling detach requests
     // ensures compatibility with operators (#2753)
+    //Remove any references to other streams
```
Comment (domoritz) [inline packages/vega-dataflow/src/EventStream.js:126]:
```suggestion
    // Remove any references to other streams
```

File: packages/vega-runtime/src/context.js [hunk 2]
```diff
@@ -65,7 +65,10 @@ Context.prototype = Subcontext.prototype = {
     // wipe out targets first for better efficiency
     const keys = Object.keys(ctx.nodes);
     for (const key of keys) ctx.nodes[key]._targets = null;
-    for (const key of keys) ctx.nodes[key].detach();
+    for (const key of keys) {
+      ctx.nodes[key].detach();
+      delete ctx.nodes[key];
```
Comment (jheer) [inline packages/vega-runtime/src/context.js:70]: Is the `delete` operation really necessary? (We null out `ctx.nodes` right after this.) Or is there a remaining pointer to that object elsewhere? If so, we might want address that directly...

If it does make an improvement to delete, we might instead assign `undefined`, as this is faster than actually removing the property.

Comment (krisdages) [reply packages/vega-runtime/src/context.js:70]: Sorry, I can't remember... I think that there may have been references to it held by forked subcontexts, but I could be wrong.

Comment (krisdages) [reply packages/vega-runtime/src/context.js:70]: I tried removing this change and rechecking the issue in my dev environment and it doesn't look to be necessary.

**Comment by domoritz** [comment]
@krisdages do you have time to wrap up this pull request so we can merge it?

File: packages/vega-transforms/src/Facet.js [hunk 3]
```diff
@@ -60,6 +60,11 @@ inherits(Facet, Transform, {
         delete flows[key];
       }
     }
+    //Remove targets that are no longer active from the active targets array.
```
Comment (domoritz) [inline packages/vega-transforms/src/Facet.js:63]:
```suggestion
    // Remove targets that are no longer active from the active targets array.
```

**Comment by krisdages** [comment]
Sorry, things have been busy, hadn't checked github in a while...

I updated the branch with the requested change and merged the latest master in, wasn't sure if you'd prefer a rebase instead...
I got an error trying to run `yarn test` locally with both master and my branch after the merge:

$ TZ=America/Los_Angeles tape 'test/**/*-test.js'
/.../vega/node_modules/ajv/dist/core.js:129
                throw new Error(`no schema with key or ref "${schemaKeyRef}"`);
                ^

Error: no schema with key or ref "http://json-schema.org/draft-06/schema#"
    at Ajv.validate (/.../vega/node_modules/ajv/dist/core.js:129:23)
    at Ajv.validateSchema (/.../vega/node_modules/ajv/dist/core.js:240:28)
    at Ajv._addSchema (/.../vega/node_modules/ajv/dist/core.js:434:18)
    at Ajv.compile (/.../vega/node_modules/ajv/dist/core.js:140:26)
    at Object.<anonymous> (/.../vega/packages/vega/test/schema-test.js:17:28)
 
Any idea what might be causing that? The CI server doesn't seem to have had a problem with it.

**Comment by jheer** [comment]
Thanks @krisdages! Fortunately I'm not seeing that problem locally.

**Pull request merged**